src/internal/common/readers/serialization_version.go

```diff
@@ -0,0 +1,181 @@
+//
+// The current on-disk format of this field is written in big endian. The
+// highest bit denotes if the item is empty because it was deleted between the
+// time we told the storage about it and when we needed to get data for it. The
+// lowest two bytes are the version number. All other bits are reserved for
+// future use.
+//
+//	MSB 31         30             16        8       0 LSB
+//	     +----------+----+---------+--------+-------+
+//	     | del flag |   reserved   | version number |
+//	     +----------+----+---------+--------+-------+
+type persistedSerializationVersion = uint32
+
+// SerializationVersion is the in-memory size of the version number that gets
+// added to the persisted serialization version.
+//
+// Right now it's only a uint16 but we can expand it to be larger so long as the
+// expanded size doesn't clash with the flags in the high-order bits.
+type SerializationVersion uint16
+
+// DefaultSerializationVersion is the current (default) version number for all
+// services. As services evolve their storage format they should begin tracking
+// their own version numbers separate from other services.
+const DefaultSerializationVersion SerializationVersion = 1
+
+var (
+	versionFormatSize                               = int(unsafe.Sizeof(persistedSerializationVersion(0)))
+	delInFlightMask   persistedSerializationVersion = 1 << ((versionFormatSize * 8) - 1)
+)
+
+// SerializationFormat is a struct describing serialization format versions and
+// flags to add for this item.
+type SerializationFormat struct {
+	Version     SerializationVersion
+	DelInFlight bool
+}
+
+func NewVersionedBackupReader(
+	format SerializationFormat,
+	baseReaders ...io.ReadCloser,
+) (io.ReadCloser, error) {
+	if format.DelInFlight && len(baseReaders) > 0 {
+		// This is a conservative check, but we can always loosen it later on if
+		// needed. At the moment we really don't expect any data if the item was
+		// deleted.
+		return nil, clues.New("item marked deleted but has reader(s)")
+	}
+
+	formattedVersion := persistedSerializationVersion(format.Version)
+	if format.DelInFlight {
+		formattedVersion |= delInFlightMask
+	}
+
+	formattedBuf := make([]byte, versionFormatSize)
+	binary.BigEndian.PutUint32(formattedBuf, formattedVersion)
+
+	versionReader := io.NopCloser(bytes.NewReader(formattedBuf))
+
+	// Need to add readers individually because types differ.
+	allReaders := make([]io.Reader, 0, len(baseReaders)+1)
+	allReaders = append(allReaders, versionReader)
+
+	for _, r := range baseReaders {
+		allReaders = append(allReaders, r)
+	}
+
+	res := &versionedBackupReader{
+		baseReaders: append([]io.ReadCloser{versionReader}, baseReaders...),
+		combined:    io.MultiReader(allReaders...),
+	}
+
+	return res, nil
+}
+
+type versionedBackupReader struct {
+	// baseReaders is a reference to the original readers so we can close them.
+	baseReaders []io.ReadCloser
+	// combined is the reader that will return all data.
+	combined io.Reader
```

Comment:
You don't think it's worth keeping a reference to the version buffer or reader here, just in case?  I'm not sure what I'd use it for either.  Just paranoid, I guess.

Reply:
technically we do have a reference to the version reader in `baseReaders`. If we find some reason we need it going forward we can add that. For the moment though I'm not sure what we'd use it for. Kopia wrapper doesn't do any validation/inspection of the written bytes so it doesn't seem like we'd need the version format at all